File: HW7/Java/HashMap.java
import java.io.File;
import java.util.Scanner;

/**
 * Created by yue on 5/1/2015.
 */
class Node{
    String key;
    int val;
    Node next;

    Node(String key, int val, Node next){
        this.key = key;
        this.val = val;
        this.next = next;
    }
}

class Pair{
    String key;
    int val;
     Pair(String key, int val){
         this.key = key;
         this.val = val;
     }
}

class Iterator{
    int current;
    int size;
    Node[] table;
    Node temp;

    Iterator(HashMap m){
        current = -1;
        size = m.size;
        table = m.table;
    }

    boolean hasNext(){
        current++;
        size--;
        if(current == table.length){
            return false;
        }
        while(table[current] == null){
            if(size == 0)
                return false;
            current++;
            size--;
        }
        temp = table[current];
        return true;
    }

    Pair getPair(){
        if(temp != null){
            Pair pair = new Pair(temp.key, temp.val);
            temp = temp.next;
            return pair;
        }
        return null;
    }
}

public class HashMap {
    int size;
    int used;
    Node[] table;

    HashMap(int initialSize){
        size = initialSize;
        used = 0;
        table = new Node[size];
    }

    private int hash(String key){
        int sum = 0;
        for (int i = 0; i < key.length(); i++)
            sum = sum + (sum << 13) + (sum >> 17) + key.charAt(i);
        return sum & (size-1);
    }

    private void grow(){
        System.out.println("doubling size, the new size is " + size);
        Iterator iterator = new Iterator(this);

        size *= 2;
        used = 0;
        table = new Node[size];

        while(iterator.hasNext()){
            Pair pair = iterator.getPair();
            while(pair != null){
                int val = pair.val;
                String key = pair.key;
                add(key, val);
                pair = iterator.getPair();
            }
        }
    }

    public int RSHash(String str)
    {
        int b     = 378551;
        int a     = 63689;
        long hash = 0;
        for(int i = 0; i < str.length(); i++)
        {
            hash = hash * a + str.charAt(i);
            a    = a * b;
        }
        return (int)Math.abs(hash % size);
    }

    public void add(String key, int val){
        int i = hash(key);
        Node temp = table[i];
        while(temp != null){
            if(temp.key.equals(key)) {
                temp.val = val;
                return;
            }
            temp = temp.next;
        }
        table[i] = new Node(key, val, table[i]);
        //when adding NEW element, used add 1
        used++;
        if(used == size){
            grow();
        }
    }

    public Integer get(String key){
        int i = hash(key);
        Node temp = table[i];
        while(temp != null){
            if(temp.key.equals(key)){
                return temp.val;
            }
            temp = temp.next;
        }
        return null;
    }

    public static void main(String[] args) throws Exception{
        HashMap dictionary = new HashMap(50000);
        HashMap wrongDict = new HashMap(50000);

        //read in dictionary
        Scanner in = new Scanner(new File("wordsEn.txt"));
        while(in.hasNextLine()){
            String dictWord = in.nextLine();
            dictionary.add(dictWord, 0);
        }

        //read in book
        in = new Scanner(new File("CompletWorksOfShakespeare.txt"));
        while(in.hasNextLine()){
            String line = in.nextLine();
            String[] words = line.split("[\\s\\.\\,\\;\\:]+");
            for(String word:words){
                if(word.equals(""))
                    continue;
                String lowerCaseWord = word.toLowerCase();
                Integer val = dictionary.get(lowerCaseWord);//case-insensitive
                if(val != null){
                    dictionary.add(lowerCaseWord, val+1);
                }
                else{
                    Integer val2 = wrongDict.get(lowerCaseWord);
                    if(val2 != null){
                        wrongDict.add(lowerCaseWord, val2 + 1);
                    }
                    else{
                        wrongDict.add(lowerCaseWord, 1);
                    }
                }
            }
        }

        //get top 100 words and its frequency
        int count = 0;
        int[] vals = new int[100];
        String[] keys = new String[100];
        Iterator iterator = new Iterator(dictionary);
        while(iterator.hasNext()){
            Pair pair = iterator.getPair();
            while(pair != null){
                count++;
                int val = pair.val;
                String key = pair.key;
                //insertion sort
                int idx = vals.length - 1;
                if(vals[idx] < val){
                    while(idx >= 1 && vals[idx-1] < val){
                        vals[idx] = vals[idx-1];
                        keys[idx] = keys[idx-1];
                        idx--;
                    }
                    vals[idx] = val;
                    keys[idx] = key;
                }
                pair = iterator.getPair();
            }
        }

        //display number of words and wrongWords
        System.out.printf("There are %d words in the dictionary\n", dictionary.used);
        System.out.printf("There are %d mis-spelled words in the book\n", wrongDict.used);

        //display top 10 words
        System.out.println("--------Top 10 words (case-insensitive)--------");
        for(int i = 0; i < 10; i++){
            System.out.println(keys[i] + "==>" + vals[i]);
        }
    }
}

File: HW7/C++/HashMap.cpp
#include <cstdint> // C++11 Required
#include <iostream>
#include<string>
using namespace std;

class HashMap {
private:
	uint32_t size;
	uint32_t used;
	struct Node {
		string key;
		int    val;
		//    Node() : key() {}
	};
	Node* table;
	void grow() {
	}
	size_t hash(const string& s) {
		uint32_t sum = 0;
		for (int i = 0; i < s.size(); i++)
			sum = sum + (sum << 13) + (sum >> 17) + s[i];
		return sum & (size - 1);
	}
public:
	HashMap(size_t initialSize)
		: size(initialSize), used(0) {
		table = new Node[size];
	}
	HashMap(const HashMap& orig);
	HashMap& operator =(const HashMap& orig);
	~HashMap() {
		delete[] table;
	}

	// if key is already there replace the value
	void add(const string& key, int val) {
		uint32_t i = hash(key);
		while (table[i].key.length() != 0) {
			if (table[i].key == key) {
				table[i].val = val;
				return;
			}
			i = (i + 1) & (size - 1);
		}
		table[i].key = key;
		table[i].val = val;
	}
	int* get(const string& key) {
		uint32_t i = hash(key);
		while (table[i].key.length() != 0) {
			if (table[i].key == key) {
				return &table[i].val;
			}
			i = (i + 1) & (size - 1);
		}
		return nullptr;
	}

	class Iterator {
	private:
		Node* current; // if the hashmap grows, this is GARBAGE
		size_t size;
	public:
		Iterator(HashMap& m) {
			current = &m.table[-1];
			size = m.size;
		}

		bool hasNext() {
			current++;
			size--;
			while (current->key.length() == 0) {
				if (size == 0)
					return false;
				current++;
				size--;
			}
			return true;
		}
		string getKey() {
			return current->key;
		}
		int getValue() {
			return current->val;
		}
	};
	friend Iterator;

};

int main() {
	HashMap m(4);
	m.add("IBM", 106);
	m.add("BABA", 88);
	int* val = m.get("IBM");
	if (val != nullptr) {
		cout << *val << endl;
	}

	HashMap::Iterator i(m);
	while (i.hasNext()) {
		cout << i.getKey() << "==>" << i.getValue() << '\n';
	}
	system("pause");
}